src/Futhark/Optimise/ReduceDeviceSyncs/MigrationTable.hs

```diff
@@ -0,0 +1,1622 @@
+    BasicOp (ArrayLit arr t)
+      | isScalar t,
+        any (isJust . subExpVar) arr ->
+          -- Migrating an array literal with free variables saves a write for
+          -- every scalar it contains. Under CUDA each scalar is written using
+          -- a synchronous write. Under OpenCL scalar constants will be written
+          -- asynchronously but free variables are still transferred using
+          -- blocking writes. If all scalars are constants then the compiler
+          -- generates more efficient code that copies static device memory.
+          graphAutoMove (one bs)
+    BasicOp UnOp {} -> graphSimple bs e
+    BasicOp BinOp {} -> graphSimple bs e
+    BasicOp CmpOp {} -> graphSimple bs e
+    BasicOp ConvOp {} -> graphSimple bs e
+    BasicOp Assert {} ->
+      -- == OpenCL =============================================================
+      --
+      -- The next read after the execution of a kernel containing an assertion
+      -- will be made asynchronous, followed by an asynchronous read to check
+      -- if any assertion failed. The runtime will then block for all enqueued
+      -- operations to finish.
+      --
+      -- Since an assertion only binds a certificate of unit type, an assertion
+      -- cannot increase the number of (read) synchronizations that occur. In
+      -- this regard it is free to migrate. The synchronization that does occur
+      -- is however (presumably) more expensive as the pipeline of GPU work will
+      -- be flushed.
+      --
+      -- Since this cost is difficult to quantify and amortize over assertion
+      -- migration candidates (cost depends on ordering of kernels and reads) we
+      -- assume it is insignificant. This will likely hold for a system where
+      -- multiple threads or processes schedules GPU work, as system-wide
+      -- throughput only will decrease if the GPU utilization decreases as a
+      -- result.
+      --
+      -- == CUDA ===============================================================
+      --
+      -- Under the CUDA backend every read is synchronous and is followed by
+      -- a full synchronization that blocks for all enqueued operations to
+      -- finish. If any enqueued kernel contained an assertion, another
+      -- synchronous read is then made to check if an assertion failed.
+      --
+      -- Migrating an assertion to save a read may thus introduce new reads, and
+      -- the total number of reads can hence either decrease, remain the same,
+      -- or even increase, subject to the ordering of reads and kernels that
+      -- perform assertions.
+      --
+      -- Since it is possible to implement the same failure checking scheme as
+      -- OpenCL using asynchronous reads (and doing so would be a good idea!)
+      -- we consider this to be acceptable.
+      --
+      -- TODO: Implement the OpenCL failure checking scheme under CUDA. This
+      --       should reduce the number of synchronizations per read to one.
```

Comment:
@athas: The differences between how reads are handled between OpenCL and CUDA are also documented here. It sounds like you might be looking at the `TODO`.

Reply:
Yes, that's fixed on master.

Reply:
Then I guess the CUDA section of the comment and the headers just can be removed, since the OpenCL section then applies to both backends. Are we maintaining other backends where these changes are relevant for? Alternatively the header can be modified.